### PR TITLE
New: Delete Quota Association

### PR DESCRIPTION
**Pull request by steveholmes**
This new functionality allows users to delete existing Quota Associations

https://uktrade.atlassian.net/browse/TARIFFS-552
https://uktrade.atlassian.net/browse/TARIFFS-562
https://uktrade.atlassian.net/browse/TARIFFS-557
https://uktrade.atlassian.net/browse/TARIFFS-568
https://uktrade.atlassian.net/browse/TARIFFS-563

### CODE REVIEW

File: app/forms/workbasket_forms/delete_quota_association_form.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        association = QuotaAssociation.find(
+          main_quota_definition_sid: @workbasket.settings.main_quota_definition_sid,
+          sub_quota_definition_sid: @workbasket.settings.sub_quota_definition_sid
+        )
+
+        ::WorkbasketValueObjects::Shared::SystemOpsAssigner.new(
+          association, system_ops.merge(operation: "D")
+        ).assign!(false)
+
+        association.destroy
```
Comment (KyleMacPherson) [inline app/forms/workbasket_forms/delete_quota_association_form.rb]: I don't think we should be destroying this until we get a success message from CDS

File: app/decorators/workbaskets/delete_quota_association_settings_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+              and sub_quota_definition_sid = :sub_quota_definition_sid
+              and (operation = 'U' or operation = 'C')
+              order by oid desc
+              limit 1
+    SQL
+
+    @deleted_quota_association = Sequel::Model.db.fetch(find_deleted_quota_association_sql,
+                           main_quota_definition_sid: main_quota_definition_sid,
+                           sub_quota_definition_sid: sub_quota_definition_sid).first
+    end
```
Comment (KyleMacPherson) [inline app/decorators/workbaskets/delete_quota_association_settings_decorator.rb]: If we don't destroy the association until we get the success message from CDS then I don't think we need this

File: app/models/quota_association.rb [hunk 3]
```diff
@@ -16,6 +16,16 @@ class QuotaAssociation < Sequel::Model
               key: :sub_quota_definition_sid,
               class: QuotaDefinition
 
+  def self.undo_deletion_by_workbasket!(workbasket_id:)
```
Comment (KyleMacPherson) [inline app/models/quota_association.rb]: What is this doing? undo deletion sounds like you are recreating a quota association but that doesn't seem to be what we are doing here?

File: app/models/workbaskets/workbasket.rb [hunk 4]
```diff
@@ -392,6 +397,10 @@ def clean_up_drafts!
             settings.collection.each(&:delete)
           end
 
+          def clean_up_quota_association_deletions!
```
Comment (KyleMacPherson) [inline app/models/workbaskets/workbasket.rb]: I'm not clear what is happening here

File: app/views/workbaskets/delete_quota_association/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,140 @@
+<% content_for :content do %>
+<!--  <div id="vue-create-quota-association-form">-->
```
Comment (KyleMacPherson) [inline app/views/workbaskets/delete_quota_association/new.html.erb]: Guess we can remove this?

File: app/views/workbaskets/delete_quota_association/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                  data-disable-with='Processing'>
+            Submit for Cross-Check
+          </button>
+
+          <%= link_to "Cancel", quota_associations_path, class: "secondary-button" %>
+        </div>
+
+      </div>
+    <%# end %>
+<!--  </div>-->
```
Comment (KyleMacPherson) [inline app/views/workbaskets/delete_quota_association/new.html.erb]: Think we can delete this

File: app/views/workbaskets/delete_quota_association/show.html.slim [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  h1.heading-large
+    | View quota association
+
+  .view-workbasket-container
+    = render "workbaskets/delete_quota_association/workflow_screens_parts/notifications/block", screen_type: :view
+    = render "workbaskets/delete_quota_association/workflow_screens_parts/actions_allowed"
+    = render "workbaskets/delete_quota_association/workflow_screens_parts/workbasket_details"
+
+    - if iam_workbasket_author? && workbasket_rejected?
+      = link_to "XXXXX quota association", '#',
```
Comment (KyleMacPherson) [inline app/views/workbaskets/delete_quota_association/show.html.slim]: I guess 'xxxxx' is a placeholder than needs to be replaced?

File: app/views/workbaskets/delete_quota_association/submitted_for_cross_check.html.slim [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+h3.heading-medium.m-t-100
+  | Next steps
+
+ul class="list"
+  li
+    = link_to "Withdraw submission", "#",
+            data: {target_url: workbasket_edit_link(workbasket.decorate), target_modal: workbasket.id},
+            class: "js-main-menu-show-withdraw-confirmation-link"
+  li
+    = link_to "View this quota association", delete_quota_association_path(workbasket.id)
```
Comment (KyleMacPherson) [inline app/views/workbaskets/delete_quota_association/submitted_for_cross_check.html.slim]: Is there a difference between this and create_quota_association_path?

File: app/views/workbaskets/delete_quota_association/workflow_screens_parts/_actions_allowed.html.slim [hunk 9]
```diff
@@ -0,0 +1,10 @@
+.view-workbasket-actions-allowed
+  - if iam_workbasket_author? && workbasket.can_withdraw?
+    = link_to "Withdraw workbasket from workflow", "#",
+    data: { target_url: withdraw_workbasket_from_workflow_delete_quota_association_url(workbasket.id), target_modal: workbasket.id },
```
Comment (KyleMacPherson) [inline app/views/workbaskets/delete_quota_association/workflow_screens_parts/_actions_allowed.html.slim]: Will we actually be able to withdraw this? I would have thought not as the object is already deleted?

File: app/forms/workbasket_forms/delete_quota_association_form.rb (hunk 1 above)
Comment (steveholmes) [reply app/forms/workbasket_forms/delete_quota_association_form.rb]: But wouldn't that go against the pattern used throughout the rest of the app?
Our normal pattern is that we add a line in *_oplog and use it's status to track it's progress through workflow and CDS?

File: app/views/workbaskets/delete_quota_association/show.html.slim (hunk 7 above)
Comment (steveholmes) [reply app/views/workbaskets/delete_quota_association/show.html.slim]: Thanks, this block should have been deleted.

File: app/views/workbaskets/delete_quota_association/submitted_for_cross_check.html.slim (hunk 8 above)
Comment (steveholmes) [reply app/views/workbaskets/delete_quota_association/submitted_for_cross_check.html.slim]: Yes this is creating a workbasket for the action of "Deleting an association"

File: app/views/workbaskets/delete_quota_association/workflow_screens_parts/_actions_allowed.html.slim (hunk 9 above)
Comment (steveholmes) [reply app/views/workbaskets/delete_quota_association/workflow_screens_parts/_actions_allowed.html.slim]: Should be fine now as not deleting

**Pull request merged**